### PR TITLE
CSS namespaces

### PR DESCRIPTION
**Pull request by 2biazdk**

### CODE REVIEW

**Comment by kasperisager** [comment]
Will review when passing.

**Comment by NiclasHedam** [comment]
There is a conflict now, @2biazdk.

File: packages/alfa-dom/src/get-attribute.ts [hunk 1]
```diff
@@ -17,43 +39,137 @@ const attributeMaps: WeakMap<Element, Map<string, string>> = new WeakMap();
...
   name: string,
-  options: Readonly<{
-    trim?: boolean;
-    lowerCase?: boolean;
-  }> = {}
+  selectorNamespace?: Namespace | string | null | AttributeOptions,
+  options: AttributeOptions = {}
 ): string | null {
-  let attributeMap = attributeMaps.get(element);
+  if (selectorNamespace !== null && typeof selectorNamespace === "object") {
```
Comment (2biazdk) [inline packages/alfa-dom/src/get-attribute.ts]: Is there an easier/prettier way of figuring out whether or not it is of an AttributeOptions interface?

File: packages/alfa-css/src/grammars/selector.ts [hunk 2]
```diff
@@ -28,7 +28,67 @@ import {
...
+export type ComplexSelector = SimpleSelector | CompoundSelector;
+
+export interface RelativeSelector {
+  readonly type: SelectorType.RelativeSelector;
+  readonly combinator: SelectorCombinator;
+  readonly left: ComplexSelector | RelativeSelector;
+  readonly right: ComplexSelector;
+}
+
+export type Selector = ComplexSelector | RelativeSelector;
```
Comment (kasperisager) [inline packages/alfa-css/src/grammars/selector.ts]: These are already defined in `types.ts`; looks like a merge gone awry.

File: packages/alfa-dom/src/get-attribute.ts [hunk 3]
```diff
@@ -1,6 +1,28 @@
...
+export function getAttribute(
+  element: Element,
+  qualifiedName: string,
+  options?: AttributeOptions
+): string | null;
+
+export function getAttribute(
+  element: Element,
+  localName: string,
+  namespace?: Namespace | string | null,
```
Comment (kasperisager) [inline packages/alfa-dom/src/get-attribute.ts]: Should be:

```ts
namespace?: Namespace | "*" | null
```

File: packages/alfa-css/src/grammars/selector.ts [hunk 4]
```diff
@@ -162,13 +128,18 @@ function typeSelector(
 }
 
 function attributeSelector(stream: Stream<Token>): AttributeSelector | null {
+  const token = stream.peek(0);
   let next = stream.next();
```
Comment (kasperisager) [inline packages/alfa-css/src/grammars/selector.ts]: Any reason for having both `token` and `next`?

File: packages/alfa-css/src/grammars/selector.ts [hunk 5]
```diff
@@ -509,6 +478,72 @@ function combineSelectors(
   return null;
 }
 
+function namespacePrefix(
```
Comment (kasperisager) [inline packages/alfa-css/src/grammars/selector.ts]:
```ts
/**
 * @see https://www.w3.org/TR/selectors/#typedef-ns-prefix
 */
```

File: packages/alfa-dom/src/matches.ts [hunk 6]
```diff
@@ -259,25 +265,49 @@ const whitespace = /\s+/;
...
+
+  if (selector.namespace !== null) {
+    // Match all namespaces
+    if (selector.namespace === "*") {
+      namespaceURI = "*";
+      // Abort when no namespace is declared
+    } else if (options.namespaces === undefined) {
+      return false;
+      // Selector namespace must match a declared namespace
+    } else if (selector.namespace !== "") {
```
Comment (kasperisager) [inline packages/alfa-dom/src/matches.ts]: No need to an `else` after a `return`.

File: packages/alfa-dom/src/matches.ts [hunk 7]
```diff
@@ -259,25 +265,49 @@ const whitespace = /\s+/;
...
+      const declaredNamespace = options.namespaces.get(selector.namespace);
+      if (declaredNamespace === undefined) {
+        return false;
+      }
+      namespaceURI = declaredNamespace;
+    }
   }
 
+  const value = getAttribute(element, selector.name, namespaceURI, {
+    lowerCase: (selector.modifier & AttributeModifier.CaseInsensitive) > 0
```
Comment (kasperisager) [inline packages/alfa-dom/src/matches.ts]:
```ts
!== 0
```

File: packages/alfa-dom/test/get-attribute.spec.tsx [hunk 8]
```diff
@@ -30,3 +31,44 @@ test("Gets an attribute value when it is defined and lowercase=true", t => {
     "foobar"
   );
 });
+
+test("Gets an attribute with a HTML namespace", t => {
+  const div = <div aria-labelledby="foobar" />;
+
+  t.equal(getAttribute(div, "aria-labelledby", Namespace.HTML), "foobar");
```
Comment (kasperisager) [inline packages/alfa-dom/test/get-attribute.spec.tsx]: `getAttribute` should return `null` in the above case; in HTML5, attributes are not assigned no namespaces, not even the HTML namespace.

File: packages/alfa-dom/test/get-attribute.spec.tsx [hunk 9]
```diff
@@ -30,3 +31,44 @@ test("Gets an attribute value when it is defined and lowercase=true", t => {
...
+});
+
+test("Gets an attribute with an SVG namespace", t => {
+  const svg: jsx.JSX.Element = {
+    nodeType: 1,
+    prefix: "svg",
```
Comment (kasperisager) [inline packages/alfa-dom/test/get-attribute.spec.tsx]: Should likely be `null`.

File: packages/alfa-dom/src/get-attribute.ts [hunk 10]
```diff
@@ -1,6 +1,35 @@
...
+  namespace: Namespace | null,
+  options?: AttributeOptions
+): string | null;
+
+export function getAttribute(
+  element: Element,
+  localName: string,
+  namespace: "*",
+  options?: AttributeOptions
+): string | Array<string> | null;
```
Comment (kasperisager) [inline packages/alfa-dom/src/get-attribute.ts]: The last overload should return `Array<string> | null`.

File: packages/alfa-dom/src/matches.ts [hunk 11]
```diff
@@ -259,25 +265,65 @@ const whitespace = /\s+/;
...
-    lowerCase: (selector.modifier & AttributeModifier.CaseInsensitive) > 0
-  });
+  if (!matchesAttributeNamespace(element, context, selector, options)) {
+    return false;
+  }
 
-  if (selector.value === null) {
-    return value !== null;
+  let value = null;
+  const option = {
```
Comment (kasperisager) [inline packages/alfa-dom/src/matches.ts]: I'd rename this `attributeOptions`.

File: packages/alfa-dom/src/get-attribute.ts [hunk 12]
```diff
@@ -17,36 +46,144 @@ const attributeMaps: WeakMap<Element, Map<string, string>> = new WeakMap();
...
+    const { localName } = attribute;
+
+    let attributeArray = attributeMap.get(localName);
+
+    if (attributeArray === undefined) {
+      attributeArray = [];
+    }
+
+    attributeArray.push(attribute);
+    attributeMap.set(localName, attributeArray);
```
Comment (kasperisager) [inline packages/alfa-dom/src/get-attribute.ts]: This line can be moved to just after line 174 to ensure that the map is only touched when initializing the array.

**Pull request merged**